source/common/upstream/load_balancer_impl.cc

```diff
@@ -615,41 +631,150 @@ bool LoadBalancerBase::isHostSetInPanic(const HostSet& host_set) const {
+    if (!local_at_end) {
+      local_hosts_in_locality = local_it->size();
+      if (local_hosts_in_locality > 0) {
+        local_locality = local_it->begin()->get()->locality();
+      }
+    }
+    if (!upstream_at_end) {
+      upstream_hosts_in_locality = upstream_it->size();
+      if (upstream_hosts_in_locality > 0) {
+        upstream_locality = upstream_it->begin()->get()->locality();
+      }
+    }
+
+    bool consume_local_locality = false;
+    bool consume_upstream_locality = false;
+    bool is_empty_consumption =
+        false; // indicates that nothing should be pushed into the percentage list
+    if (!local_at_end && upstream_at_end) { // Rule 1
+      consume_local_locality = true;
+    } else if (local_at_end && !upstream_at_end) { // Rule 1
+      consume_upstream_locality = true;
+    } else if (local_hosts_in_locality == 0) { // Rule 2
+      consume_local_locality = true;
+      is_empty_consumption = true;
+    } else if (upstream_hosts_in_locality == 0) { // Rule 2
+      consume_upstream_locality = true;
+      is_empty_consumption = true;
+    } else if (LocalityEqualTo()(local_locality, upstream_locality)) { // Rule 3
+      consume_local_locality = true;
+      consume_upstream_locality = true;
+    } else if (is_current_local_locality) { // Rule 4
+      consume_local_locality = true;
+    } else if (LocalityLess()(local_locality, upstream_locality)) { // Rule 5
+      consume_local_locality = true;
+    } else { // Rule 5: upstream_locality < local_locality
+      consume_upstream_locality = true;
+    }
 
-  size_t i = 0;
-  for (const auto& locality_hosts : hosts_per_locality.get()) {
-    ret[i++] = total_hosts > 0 ? 10000ULL * locality_hosts.size() / total_hosts : 0;
+    if (consume_local_locality) {
+      if (!is_empty_consumption) {
+        // Note that local_hosts_in_locality > 0 is always true, but the check is added for extra
+        // defense
+        local_percentage.emplace_back(local_hosts_in_locality > 0
+                                          ? 10000ULL * local_hosts_in_locality / total_local_hosts
+                                          : 0);
+      }
+      ++local_it;
+    } else if (!is_empty_consumption) {
+      local_percentage.emplace_back(0);
+    }
+    if (consume_upstream_locality) {
+      if (!is_empty_consumption) {
+        // Note that upstream_hosts_in_locality > 0 is always true, but the check is added for extra
+        // defense
+        upstream_percentage.emplace_back(upstream_hosts_in_locality > 0
+                                             ? 10000ULL * upstream_hosts_in_locality /
+                                                   total_upstream_hosts
+                                             : 0);
+        upstream_to_percentage_index[upstream_index] = upstream_percentage.size() - 1;
+      }
+      ++upstream_it;
+      ++upstream_index;
+    } else if (!is_empty_consumption) {
+      upstream_percentage.emplace_back(0);
+    }
```

Comment:
Basically I know what you want. But this code is pretty hard to follow or maintian. Could we make it more simple and clear?

Because all traffic will be routed to the upstream. So we only need to calculate the locality percentage of upstream cluster and related local locality percentage. Finally, we can output a structure like this:

```
percentages:
- upstream_percentage: xxxx
  local_percentage: xxxx
- upstream_percentage: xxxx
  local_percentage: xxxx

```

Reply:
used your suggestions to make it much simpler using a map